plugins.d/alarm-notify.sh.in

```diff
@@ -859,6 +554,51 @@ if [ "${SEND_AWSSNS}" = "YES" -a -z "${aws}" ]
     fi
 fi
 
+# -----------------------------------------------------------------------------
+# find the recipients' addresses per method
+
+# netdata may call us with multiple roles, and roles may have multiple but
+# overlapping recipients - so, here we find the unique recipients.
+for method_name in ${method_names} ; do
+    send="SEND_${method_name^^}"
+    if [ ${!send} = "NO" ] ; then
+        continue
+    fi
+
+    declare -A arr=()
+
+    for x in ${roles//,/ } ; do
+        # the roles 'silent' and 'disabled' mean:
+        # don't send a notification for this role
+        [ "${x}" = "silent" -o "${x}" = "disabled" ] && continue
+
+        role_recipients="role_recipients_${method_name}"
+        role_recipients=${!role_recipients}
+        default_recipient="DEFAULT_RECIPIENT_${method_name^^}"
+
+        a="${role_recipients[${x}]}"
+        [ -z "${a}" ] && a="${!default_recipient}"
+        for r in ${a//,/ } ; do
+            [ "${r}" != "disabled" ] && filter_recipient_by_criticality ${method_name} "${r}" && arr[${r/|*/}]="1"
+        done
+    done
+
+    # build the list of recipients
+    to="to_${method_name}"
+    if [ ${method_name} = "email" ] ; then
+        declare to_${method_name}
+        for x in "${!arr[@]}" ; do
+            [ ! -z "${!to}" ] && to_email="${!to}, "
+            to_email="${!to}${x}"
```

Comment:
`to_email` ?

Reply:
That's part of the special-case for email handling.  We ask for email addresses separated by a comma and a space, not just a space, so we need to parse the recipient list differently.

Somewhat ironically, the line right above the first context line shown here is the one that indicates this.

Reply:
hm... then it should be probably better to run the loop without that `if` and after the loop do:

```sh
# remove multiple adjacent spaces from ${to_email}
to_email="${to_email//  / }"
# remove leading space
to_email="${to_email/# /}"
# remove trailing space
to_email="${to_email/% /}"

# emails have to be separated by comma and a space
to_email="${to_email// /, }"
```

or even (the following is probably better compared to the previous one - it handles multi-space separated values better):

```sh
fix_to_email() {
  to_email=
  while [ ! -z "${1}" ]
  do
    [ ! -z "${to_email}" ] && to_email="${to_email}, "
    to_email="${to_email}${1}"
    shift 1
  done
}

# ${to_email} without quotes here
fix_to_email ${to_email}
```

Reply:
Or, alternatively, we could push that logic into the function that actually sends the email and just make the API for send functions more consistent, though I'm not quite sure exactly how hard that may be (the diff might be huge, because it might require relocating the template email into the function body).

Reply:
yes, I like that too.
Though I think email is a bit out of scope at the end of the file...